File: tests/Aspire.Hosting.CodeGeneration.Elixir.ExTests/test/aspire_watch_test.exs
defmodule Aspire.WatchTest do
  @moduledoc """
  Tests for `watch.exs`, the AppHost file watcher.

  Each test starts a real `elixir` process, so the timeouts stay short and every process is
  stopped in `on_exit/1`.
  """

  use ExUnit.Case, async: false

  @watch_script Path.expand(
                  "../../../src/Aspire.Hosting.CodeGeneration.Elixir/Resources/watch.exs",
                  __DIR__
                )

  @timeout 10_000

  test "watch restarts child when apphost.exs mtime changes" do
    directory = temporary_directory()

    write(directory, "apphost.exs", """
    IO.puts("APPHOST-STARTED")
    Process.sleep(60_000)
    """)

    port = start_watch(directory, ["apphost.exs"])

    started = await_output(port, "APPHOST-STARTED", @timeout)
    assert started =~ "APPHOST-STARTED"

    # A second granularity mtime needs a stamp that is clearly newer than the snapshot.
    File.touch!(Path.join(directory, "apphost.exs"), System.os_time(:second) + 2)

    output = await_output(port, "APPHOST-STARTED", @timeout, 2, started)
    assert count(output, "APPHOST-STARTED") >= 2
    assert output =~ "[aspire-watch] restarting: apphost.exs"
  end

  test "watch ignores _build and the generated SDK" do
    directory = temporary_directory()

    write(directory, "apphost.exs", """
    IO.puts("APPHOST-STARTED")
    Process.sleep(60_000)
    """)

    File.mkdir_p!(Path.join(directory, "_build"))
    File.mkdir_p!(Path.join([directory, ".aspire", "modules"]))
    write(directory, "_build/stale.ex", "x = 1\n")
    write(directory, ".aspire/modules/aspire_generated.ex", "y = 1\n")

    port = start_watch(directory, ["apphost.exs"])
    started = await_output(port, "APPHOST-STARTED", @timeout)
    assert started =~ "APPHOST-STARTED"

    stamp = System.os_time(:second) + 2
    File.touch!(Path.join(directory, "_build/stale.ex"), stamp)
    File.touch!(Path.join(directory, ".aspire/modules/aspire_generated.ex"), stamp)

    output = collect(port, 2_000, started)
    refute output =~ "[aspire-watch] restarting:"
    assert count(output, "APPHOST-STARTED") == 1
  end

  test "watch forwards child exit code" do
    directory = temporary_directory()

    write(directory, "apphost.exs", """
    IO.puts("APPHOST-FAILED")
    System.halt(3)
    """)

    port = start_watch(directory, ["apphost.exs", "--once"])

    assert {output, 3} = await_exit(port, @timeout)
    assert output =~ "APPHOST-FAILED"
  end

  test "watch keeps waiting after the apphost stops on its own" do
    directory = temporary_directory()

    write(directory, "apphost.exs", """
    IO.puts("APPHOST-STARTED")
    System.halt(3)
    """)

    port = start_watch(directory, ["apphost.exs"])

    output = await_output(port, "Waiting for a file change", @timeout)
    assert output =~ "[aspire-watch] apphost stopped with status 3. Waiting for a file change."
    assert Port.info(port) != nil
  end

  # ── Helpers ───────────────────────────────────────────────────────────────

  defp start_watch(directory, arguments) do
    port =
      Port.open(
        {:spawn_executable, System.find_executable("elixir")},
        [
          :binary,
          :exit_status,
          :hide,
          :stderr_to_stdout,
          cd: directory,
          args: [@watch_script | arguments]
        ]
      )

    on_exit(fn -> stop(port) end)
    port
  end

  defp stop(port) do
    case Port.info(port, :os_pid) do
      {:os_pid, os_pid} ->
        _ = System.cmd("kill", ["-KILL", Integer.to_string(os_pid)], stderr_to_stdout: true)
        :ok

      nil ->
        :ok
    end
  end

  # Reads until the marker appeared `wanted` times, or the timeout elapsed.
  defp await_output(port, marker, timeout, wanted \\ 1, acc \\ "") do
    if count(acc, marker) >= wanted do
      acc
    else
      receive do
        {^port, {:data, data}} -> await_output(port, marker, timeout, wanted, acc <> data)
        {^port, {:exit_status, _status}} -> acc
      after
        timeout ->
          flunk("no #{inspect(marker)} (x#{wanted}) from the watcher in #{timeout}ms: #{acc}")
      end
    end
  end

  defp await_exit(port, timeout, acc \\ "") do
    receive do
      {^port, {:data, data}} -> await_exit(port, timeout, acc <> data)
      {^port, {:exit_status, status}} -> {acc, status}
    after
      timeout -> flunk("the watcher did not stop in #{timeout}ms: #{acc}")
    end
  end

  defp collect(port, timeout, acc) do
    receive do
      {^port, {:data, data}} -> collect(port, timeout, acc <> data)
      {^port, {:exit_status, _status}} -> acc
    after
      timeout -> acc
    end
  end

  defp count(text, marker) do
    text |> String.split(marker) |> length() |> Kernel.-(1)
  end

  defp temporary_directory do
    path =
      Path.join(
        System.tmp_dir!(),
        "aspire_watch_#{System.unique_integer([:positive])}"
      )

    File.mkdir_p!(path)
    on_exit(fn -> File.rm_rf(path) end)
    path
  end

  defp write(directory, name, content) do
    path = Path.join(directory, name)
    File.mkdir_p!(Path.dirname(path))
    File.write!(path, content)
    path
  end
end

File: tests/Aspire.Hosting.CodeGeneration.Dart.DartTests/test/aspire_watch_test.dart
// Tests for `watch.dart`, the AppHost file watcher.
//
// Each test starts a real `dart` process, so the timeouts stay short and every
// process is stopped in `addTearDown`.

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';

/// The watcher the code generator copies into `.aspire/modules/`. `dart test`
/// runs from the package directory, so the path is relative to it.
final String watchScript = File(
  '../../src/Aspire.Hosting.CodeGeneration.Dart/Resources/watch.dart',
).absolute.path;

const Duration timeout = Duration(seconds: 10);

void main() {
  test('watch restarts child when apphost.dart mtime changes', () async {
    final Directory directory = temporaryDirectory();
    write(
      directory,
      'pubspec.yaml',
      'name: watch_app\npublish_to: none\n\nenvironment:\n  sdk: ^3.8.0\n',
    );
    write(directory, 'apphost.dart', '''
import 'dart:io';

Future<void> main(List<String> args) async {
  stdout.writeln('APPHOST-STARTED');
  await stdout.flush();
  await Future<void>.delayed(const Duration(seconds: 60));
}
''');

    final _Watcher watcher = await startWatch(directory, <String>[
      'apphost.dart',
    ]);

    await watcher.awaitOutput('APPHOST-STARTED', 1);

    // The watcher compares modification times, so the stamp has to be clearly
    // newer than the snapshot it already took.
    File(
      '${directory.path}/apphost.dart',
    ).setLastModifiedSync(DateTime.now().add(const Duration(seconds: 2)));

    final String output = await watcher.awaitOutput('APPHOST-STARTED', 2);
    expect(output, contains('[aspire-watch] restarting: apphost.dart'));
  });

  test('watch forwards child exit code with --once', () async {
    final Directory directory = temporaryDirectory();
    write(
      directory,
      'pubspec.yaml',
      'name: watch_app\npublish_to: none\n\nenvironment:\n  sdk: ^3.8.0\n',
    );
    write(directory, 'apphost.dart', '''
import 'dart:io';

void main() {
  stdout.writeln('APPHOST-FAILED');
  exit(3);
}
''');

    final _Watcher watcher = await startWatch(directory, <String>[
      'apphost.dart',
      '--once',
    ]);

    final int status = await watcher.process.exitCode.timeout(timeout);
    expect(status, 3);
    expect(watcher.output, contains('APPHOST-FAILED'));
  });
}

// ── Helpers ──────────────────────────────────────────────────────────────────

/// A running watcher and everything it wrote so far.
class _Watcher {
  _Watcher(this.process);

  final Process process;
  final StringBuffer _output = StringBuffer();
  final List<void Function()> _listeners = <void Function()>[];

  String get output => _output.toString();

  void collect(Stream<List<int>> stream) {
    stream.transform(utf8.decoder).listen((String chunk) {
      _output.write(chunk);
      for (final void Function() listener in _listeners.toList()) {
        listener();
      }
    });
  }

  /// Waits until [marker] appeared [wanted] times, then returns everything the
  /// watcher wrote.
  Future<String> awaitOutput(String marker, int wanted) {
    final Completer<String> completer = Completer<String>();

    void check() {
      if (!completer.isCompleted && count(output, marker) >= wanted) {
        completer.complete(output);
      }
    }

    _listeners.add(check);
    check();

    return completer.future.timeout(
      timeout,
      onTimeout: () => throw StateError(
        'no "$marker" (x$wanted) from the watcher in '
        '${timeout.inMilliseconds}ms: $output',
      ),
    );
  }
}

Future<_Watcher> startWatch(Directory directory, List<String> arguments) async {
  final Process process = await Process.start(
    Platform.resolvedExecutable,
    <String>['run', watchScript, ...arguments],
    workingDirectory: directory.path,
  );

  final _Watcher watcher = _Watcher(process);
  watcher.collect(process.stdout);
  watcher.collect(process.stderr);

  // The watcher stops its own child on SIGTERM, so the AppHost never outlives
  // the test.
  addTearDown(() async {
    process.kill(ProcessSignal.sigterm);
    await process.exitCode.timeout(
      timeout,
      onTimeout: () {
        process.kill(ProcessSignal.sigkill);
        return -1;
      },
    );
  });

  return watcher;
}

int count(String text, String marker) => text.split(marker).length - 1;

Directory temporaryDirectory() {
  final Directory directory = Directory.systemTemp.createTempSync(
    'aspire_watch_',
  );
  addTearDown(() {
    if (directory.existsSync()) {
      directory.deleteSync(recursive: true);
    }
  });
  return directory;
}

void write(Directory directory, String name, String content) {
  final File file = File('${directory.path}/$name');
  file.parent.createSync(recursive: true);
  file.writeAsStringSync(content);
}
